client/src/main/kotlin/io/spine/examples/pingh/client/LoginFlow.kt

```diff
@@ -0,0 +1,275 @@
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.pingh.client
+
+import com.google.protobuf.Duration
+import io.spine.examples.pingh.github.UserCode
+import io.spine.examples.pingh.github.Username
+import io.spine.examples.pingh.sessions.SessionId
+import io.spine.examples.pingh.sessions.buildBy
+import io.spine.examples.pingh.sessions.withSession
+import io.spine.examples.pingh.sessions.command.LogUserIn
+import io.spine.examples.pingh.sessions.command.VerifyUserLoginToGitHub
+import io.spine.examples.pingh.sessions.event.UserCodeReceived
+import io.spine.examples.pingh.sessions.event.UserIsNotLoggedIntoGitHub
+import io.spine.examples.pingh.sessions.event.UserLoggedIn
+import io.spine.net.Url
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+/**
+ * The control flow of the user login process.
+ *
+ * The flow consists of two consecutive stages. To successfully complete the login process,
+ * first enter a username and obtain a user code, then verify the login.
+ *
+ * @param client enables interaction with the Pingh server.
+ * @param session the information about the current user session.
+ */
+public class LoginFlow internal constructor(
+    private val client: DesktopClient,
+    private val session: MutableStateFlow<UserSession?>
+) {
+    /**
+     * Current stage of the login process.
+     */
+    private val stage: MutableStateFlow<LoginStage> =
+        MutableStateFlow(EnterUsername(client, session, this::moveToNextStage))
+
+    /**
+     * Returns the immutable state of the current login stage.
+     */
+    public fun currentStage(): StateFlow<LoginStage> = stage
+
+    /**
+     * Updates value of the current stage of the login process.
+     */
+    private fun moveToNextStage(nextStage: LoginStage) {
+        stage.value = nextStage
+    }
+}
+
+/**
+ * Stages of login process.
+ */
+public interface LoginStage
+
+/**
+ * Manages the stage of the login process where the user must enter their username
```

Comment:
"A stage of the login flow on which the user enters their GitHub username and receives a user code in return."